log.go

```diff
@@ -34,26 +40,93 @@ const (
 	LevelField = "level"
 	// Component is a field that represents component - e.g. service or
 	// function
-	Component = "component"
+	Component = "trace.component"
+	// ComponentFields is a fields compoonent
+	ComponentFields = "trace.fields"
+	// DefaultComponentPadding is a default char padding for component
+	DefaultComponentPadding = 11
+	// DefaultLevelPadding is a default char padding for component
+	DefaultLevelPadding = 4
 )
 
 // TextFormatter is logrus-compatible formatter and adds
 // file and line details to every logged entry.
 type TextFormatter struct {
-	log.TextFormatter
+	// DisableTimestamp disables timestamp output (useful when outputting to
+	// systemd logs)
+	DisableTimestamp bool
+	// ComponentPadding is a padding to pick when displaying
+	// and formatting component field, default is set to 11
+	ComponentPadding int
 }
 
 // Format implements logrus.Formatter interface and adds file and line
-func (tf *TextFormatter) Format(e *log.Entry) ([]byte, error) {
+func (tf *TextFormatter) Format(e *log.Entry) (data []byte, err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			data = append([]byte("panic in log formatter\n"), rundebug.Stack()...)
+			return
+		}
+	}()
+	var file string
 	if frameNo := findFrame(); frameNo != -1 {
 		t := newTrace(frameNo, nil)
-		new := e.WithFields(log.Fields{FileField: t.Loc(), FunctionField: t.FuncName()})
-		new.Time = e.Time
-		new.Level = e.Level
-		new.Message = e.Message
-		e = new
+		file = t.Loc()
+	}
+
+	w := &writer{bytes.Buffer{}}
+
+	// time
+	if !tf.DisableTimestamp {
+		w.writeField(e.Time.Format(time.RFC3339))
+	}
+
+	// level
+	w.writeField(strings.ToUpper(padMax(e.Level.String(), DefaultLevelPadding)))
+
+	// component, always output
+	componentI, ok := e.Data[Component]
+	if !ok {
+		componentI = ""
+	}
+	component, ok := componentI.(string)
+	if !ok {
+		component = fmt.Sprintf("%v", componentI)
+	}
+	padding := DefaultComponentPadding
+	if tf.ComponentPadding != 0 {
+		padding = tf.ComponentPadding
 	}
-	return (&tf.TextFormatter).Format(e)
+	if w.Len() > 0 {
+		w.WriteByte(' ')
+	}
+	if component != "" {
+		component = fmt.Sprintf("[%v]", component)
+	}
+	component = strings.ToUpper(padMax(component, padding))
+	if component[len(component)-1] != ' ' {
```

Comment:
Isn't the last character always `]` according to the Sprintf call above? So this block seems redundant because the component string already ends with `]`. Unless I'm missing something.

Reply:
yeah, it could be truncated if `padMax` truncated it to `padding`